volatility3/framework/plugins/linux/netfilter.py

```diff
@@ -0,0 +1,716 @@
+
+    @classmethod
+    @abstractmethod
+    def symtab_checks(cls, vmlinux: interfaces.context.ModuleInterface) -> bool:
+        """This method on each sublasss will be called to evaluate if the kernel
+        being analyzed fulfill the type & symbols requirements for the implementation.
+        The first class returning True will be instantiated and called via the
+        run() method.
+
+        Returns:
+            bool: True if the kernel being analyzed fulfill the class requirements.
+        """
+
+    def _proto_hook_loop(self) -> Iterator[Tuple[int, str, int, str]]:
+        """Flattens the protocol families and hooks"""
+        for proto_idx, proto in enumerate(AbstractNetfilter.PROTO_HOOKS):
+            if proto == PROTO_NOT_IMPLEMENTED:
+                continue
+            if proto.name not in self.subscribed_protocols():
+                # This protocol is not managed in this object
+                continue
+            for hook_idx, hook_name in enumerate(proto.hooks):
+                yield proto_idx, proto.name, hook_idx, hook_name
+
+    def build_nf_hook_ops_array(self, nf_hook_entries):
+        """Function helper to build the nf_hook_ops array when it is not part of the
+        struct 'nf_hook_entries' definition.
+
+        nf_hook_ops was stored adjacent in memory to the nf_hook_entry array, in the
+        new struct 'nf_hook_entries'. However, this 'nf_hooks_ops' array 'orig_ops' is
+        not part of the 'nf_hook_entries' struct. So, we need to calculate the offset.
+
+            struct nf_hook_entries {
+                u16                         num_hook_entries; /* plus padding */
+                struct nf_hook_entry        hooks[];
+                //const struct nf_hook_ops *orig_ops[];
+            }
+        """
+        nf_hook_entry_size = self.vmlinux.get_type("nf_hook_entry").size
+        orig_ops_addr = (
+            nf_hook_entries.hooks.vol.offset
+            + nf_hook_entry_size * nf_hook_entries.num_hook_entries
+        )
+        orig_ops = self._context.object(
+            object_type=self.get_symbol_fullname("array"),
+            offset=orig_ops_addr,
+            subtype=self.vmlinux.get_type("pointer"),
+            layer_name=self.layer_name,
+            count=nf_hook_entries.num_hook_entries,
+        )
+
+        return orig_ops
+
+    def subscribed_protocols(self) -> Tuple[str]:
+        """Allows to select which PROTO_HOOKS protocols will be processed by the
+        Netfiler subclass.
+        """
+
+        # Most implementation handlers respond to these protocols, except for
+        # the ingress hook, which specifically handles the 'NETDEV' protocol.
+        # However, there is no corresponding Netfilter hook implementation for
+        # the INET protocol in the kernel. AFAIU, this is used as
+        # 'NFPROTO_INET = NFPROTO_IPV4 || NFPROTO_IPV6'
+        # in other parts of the kernel source code.
+        return ("IPV4", "ARP", "BRIDGE", "IPV6", "DECNET")
```

Comment:
Feels like this should be defined as a constant at the top of the module?  Any reason this list is different?

Reply:
Each class overrides this, so I think it's convenient to keep them at hand, in its own class, for better readability

Reply:
Ok, makes sense, happy with that.  5:)